File: sdk/javascript/tests/workflows/worker.test.ts
import { mkdtemp, rm } from "node:fs/promises";
import { createServer } from "node:net";
import type { Server, Socket } from "node:net";
import { join } from "node:path";
import { afterEach, beforeEach, describe, expect, test } from "bun:test";
import { closeInternalSocketPoolsForTests } from "../../src/tako/socket";
import { WorkflowsClient } from "../../src/workflows/rpc-client";
import type { Run } from "../../src/workflows/types";
import type { RegisteredWorkflow, WorkflowHandler } from "../../src/workflows/worker";
import { Worker } from "../../src/workflows/worker";

class MockServer {
  server!: Server;
  path = "";
  private tasks: Run[] = [];
  private idCounter = 0;

  async start(dir: string): Promise<void> {
    this.path = join(dir, "srv.sock");
    this.server = createServer((socket: Socket) => this.handleConnection(socket));
    await new Promise<void>((r) => this.server.listen(this.path, r));
  }

  async close(): Promise<void> {
    await new Promise<void>((r) => this.server.close(() => r()));
  }

  seed(task: Partial<Run> & { name: string }): string {
    const id = `t${++this.idCounter}`;
    this.tasks.push({
      id,
      name: task.name,
      payload: task.payload ?? {},
      status: "pending",
      attempts: 0,
      retries: task.retries ?? 2,
      runAt: task.runAt ?? Date.now(),
      leaseUntil: null,
      workerId: null,
      lastError: null,
      stepState: task.stepState ?? {},
      createdAt: Date.now(),
      uniqueKey: null,
    });
    return id;
  }

  find(id: string): Run | undefined {
    return this.tasks.find((t) => t.id === id);
  }

  private handleConnection(socket: Socket): void {
    let buf = "";
    socket.on("data", (chunk: Buffer) => {
      buf += chunk.toString("utf8");
      let nl: number;
      while ((nl = buf.indexOf("\n")) !== -1) {
        const line = buf.slice(0, nl);
        buf = buf.slice(nl + 1);
        try {
          const cmd = JSON.parse(line) as Record<string, unknown>;
          const resp = this.dispatch(cmd);
          socket.write(`${JSON.stringify(resp)}\n`);
        } catch (err) {
          socket.write(`${JSON.stringify({ status: "error", message: String(err) })}\n`);
        }
      }
    });
  }

  private dispatch(cmd: Record<string, unknown>): unknown {
    switch (cmd["command"]) {
      case "claim_run": {
        const names = cmd["names"] as string[];
        const task = this.tasks.find(
          (t) => t.status === "pending" && names.includes(t.name) && t.runAt <= Date.now(),
        );
        if (!task) return { status: "ok", data: null };
        task.status = "running";
        task.attempts += 1;
        task.workerId = cmd["worker_id"] as string;
        return {
          status: "ok",
          data: {
            id: task.id,
            name: task.name,
            payload: task.payload,
            status: task.status,
            attempts: task.attempts,
            max_attempts: task.retries + 1,
            run_at_ms: task.runAt,
            step_state: task.stepState,
          },
        };
      }
      case "heartbeat_run":
        return { status: "ok", data: {} };
      case "save_step": {
        const task = this.find(cmd["id"] as string);
        if (task) {
          // Steps table model: append (step_name, result) per call.
          const stepName = cmd["step_name"] as string;
          task.stepState = { ...task.stepState, [stepName]: cmd["result"] };
        }
        return { status: "ok", data: {} };
      }
      case "complete_run": {
        const task = this.find(cmd["id"] as string);
        if (task) {
          task.status = "succeeded";
          task.workerId = null;
        }
        return { status: "ok", data: {} };
      }
      case "cancel_run": {
        const task = this.find(cmd["id"] as string);
        if (task) {
          task.status = "cancelled";
          task.lastError = (cmd["reason"] as string) ?? null;
          task.workerId = null;
        }
        return { status: "ok", data: {} };
      }
      case "defer_run": {
        const task = this.find(cmd["id"] as string);
        if (task) {
          task.status = "pending";
          task.runAt = (cmd["wake_at_ms"] as number) ?? Number.MAX_SAFE_INTEGER;
          task.workerId = null;
        }
        return { status: "ok", data: {} };
      }
      case "wait_for_event": {
        const task = this.find(cmd["id"] as string);
        if (task) {
          task.status = "pending";
          task.runAt = (cmd["timeout_at_ms"] as number) ?? Number.MAX_SAFE_INTEGER;
          task.workerId = null;
        }
        return { status: "ok", data: {} };
      }
      case "fail_run": {
        const task = this.find(cmd["id"] as string);
        if (task) {
          if (cmd["finalize"]) {
            task.status = "dead";
          } else {
            task.status = "pending";
            task.runAt = (cmd["next_run_at_ms"] as number) ?? Date.now();
          }
          task.lastError = cmd["error"] as string;
          task.workerId = null;
        }
        return { status: "ok", data: {} };
      }
      default:
        return { status: "error", message: `unknown: ${String(cmd["command"])}` };
    }
  }
}

function registry(handlers: Record<string, WorkflowHandler>): Map<string, RegisteredWorkflow> {
  return new Map(Object.entries(handlers).map(([name, handler]) => [name, { handler }]));
}

describe("Worker", () => {
  let dir: string;
  let mock: MockServer;
  let client: WorkflowsClient;

  beforeEach(async () => {
    closeInternalSocketPoolsForTests();
    dir = await mkdtemp(join("/tmp", "tako-worker-"));
    mock = new MockServer();
    await mock.start(dir);
    client = new WorkflowsClient(mock.path);
  });

  afterEach(async () => {
    closeInternalSocketPoolsForTests();
    await mock.close();
    await rm(dir, { recursive: true, force: true });
  });

  test("processes one task and marks it succeeded", async () => {
    const seen: unknown[] = [];
    const worker = new Worker({
      client,
      workerId: "w1",
      registry: registry({
        echo: (p) => {
          seen.push(p);
          return "ok";
        },
      }),
    });

    const id = mock.seed({ name: "echo", payload: { hello: 1 } });
    expect(await worker.processOnce()).toBe(true);
    expect(seen).toEqual([{ hello: 1 }]);
    expect(mock.find(id)?.status).toBe("succeeded");
  });

  test("processOnce returns false when nothing is eligible", async () => {
    const worker = new Worker({ client, workerId: "w1", registry: registry({}) });
    expect(await worker.processOnce()).toBe(false);
  });

  test("failing handler exhausts retries and dies", async () => {
    const worker = new Worker({
      client,
      workerId: "w1",
      baseBackoffMs: 1,
      maxBackoffMs: 2,
      registry: registry({
        flaky: () => {
          throw new Error("boom");
        },
      }),
    });

    const id = mock.seed({ name: "flaky", retries: 1 });
    await worker.processOnce();
    expect(mock.find(id)?.status).toBe("pending");
    expect(mock.find(id)?.attempts).toBe(1);

    await new Promise((r) => setTimeout(r, 10));
    await worker.processOnce();
    expect(mock.find(id)?.status).toBe("dead");
  });

  test("emits run/step lifecycle log lines", async () => {
    const writes: string[] = [];
    const originalWrite = process.stdout.write.bind(process.stdout);
    process.stdout.write = ((chunk: unknown): boolean => {
      writes.push(typeof chunk === "string" ? chunk : String(chunk));
      return true;
    }) as typeof process.stdout.write;

    try {
      const worker = new Worker({
        client,
        workerId: "w1",
        registry: registry({
          greet: async (_p, ctx) => {
            await ctx.run("fetch", () => "ok");
          },
        }),
      });
      mock.seed({ name: "greet" });
      await worker.processOnce();
    } finally {
      process.stdout.write = originalWrite;
    }

    const lines = writes
      .flatMap((c) => c.split("\n"))
      .filter((l) => l.length > 0)
      .map((l) => JSON.parse(l) as Record<string, unknown>);

    const find = (msg: string): Record<string, unknown> | undefined =>
      lines.find((l) => l["msg"] === msg);
    expect(find("Workflow started")).toMatchObject({ level: "info", scope: "worker:greet" });
    expect(find("Step completed")).toMatchObject({ level: "info", scope: "worker:greet" });
    expect(find("Step completed")!["fields"]).toMatchObject({ step: "fetch" });
    expect(find("Workflow completed")).toMatchObject({ level: "info", scope: "worker:greet" });
  });

  test("passes scoped loggers to workflow and step contexts", async () => {
    const writes: string[] = [];
    const originalWrite = process.stdout.write.bind(process.stdout);
    process.stdout.write = ((chunk: unknown): boolean => {
      writes.push(typeof chunk === "string" ? chunk : String(chunk));
      return true;
    }) as typeof process.stdout.write;

    let id = "";
    try {
      const worker = new Worker({
        client,
        workerId: "w1",
        registry: registry({
          cleanup: async (_p, ctx) => {
            expect(ctx.runId).toBe(id);
            expect(ctx.workflowName).toBe("cleanup");
            expect(ctx.attempt).toBe(1);
            ctx.logger.info("workflow cleanup");
            await ctx.run("delete-temp", (step) => {
              expect(step.runId).toBe(id);
              expect(step.workflowName).toBe("cleanup");
              expect(step.stepName).toBe("delete-temp");
              expect(step.attempt).toBe(1);
              step.logger.info("step cleanup");
              return "ok";
            });
          },
        }),
      });
      id = mock.seed({ name: "cleanup" });
      await worker.processOnce();
    } finally {
      process.stdout.write = originalWrite;
    }

    const lines = writes
      .flatMap((c) => c.split("\n"))
      .filter((l) => l.length > 0)
      .map((l) => JSON.parse(l) as Record<string, unknown>);

    const workflowLog = lines.find((l) => l["msg"] === "workflow cleanup");
    expect(workflowLog).toMatchObject({
      level: "info",
      scope: "cleanup",
      fields: { runId: id, workflow: "cleanup" },
    });

    const stepLog = lines.find((l) => l["msg"] === "step cleanup");
    expect(stepLog).toMatchObject({
      level: "info",
      scope: "cleanup:delete-temp",
      fields: { runId: id, workflow: "cleanup", step: "delete-temp" },
    });
  });

  test("emits Step cached on replay and Workflow cancelled on bail", async () => {
    const writes: string[] = [];
    const originalWrite = process.stdout.write.bind(process.stdout);
    process.stdout.write = ((chunk: unknown): boolean => {
      writes.push(typeof chunk === "string" ? chunk : String(chunk));
      return true;
    }) as typeof process.stdout.write;

    try {
      let pass = 0;
      const worker = new Worker({
        client,
        workerId: "w1",
        baseBackoffMs: 1,
        maxBackoffMs: 2,
        registry: registry({
          quit: async (_p, ctx) => {
            await ctx.run("prep", () => "v");
            pass += 1;
            if (pass === 1) throw new Error("retry");
            ctx.bail("done");
          },
        }),
      });
      mock.seed({ name: "quit", retries: 2 });
      await worker.processOnce();
      await new Promise((r) => setTimeout(r, 10));
      await worker.processOnce();
    } finally {
      process.stdout.write = originalWrite;
    }

    const lines = writes
      .flatMap((c) => c.split("\n"))
      .filter((l) => l.length > 0)
      .map((l) => JSON.parse(l) as Record<string, unknown>);
    const msgs = lines.map((l) => l["msg"]);
    expect(msgs).toContain("Step cached");
    expect(msgs).toContain("Workflow cancelled");
    const cancelled = lines.find((l) => l["msg"] === "Workflow cancelled");
    expect(cancelled).toMatchObject({ level: "info" });
    expect(cancelled!["fields"]).toMatchObject({ reason: "done" });
  });

  test("ctx.run memoizes across retries", async () => {
    const runs: Record<string, number> = { a: 0, b: 0 };
    let forceFail = true;
    const handler: WorkflowHandler = async (_payload, ctx) => {
      const v = await ctx.run("a", () => {
        runs.a += 1;
        return "user-1";
      });
      await ctx.run("b", () => {
        runs.b += 1;
        if (forceFail) throw new Error("fail-b");
        return v;
      });
    };

    const worker = new Worker({
      client,
      workerId: "w1",
      baseBackoffMs: 1,
      maxBackoffMs: 2,
      registry: registry({ multi: handler }),
    });

    const id = mock.seed({ name: "multi", retries: 4 });
    await worker.processOnce();
    expect(mock.find(id)?.status).toBe("pending");
    expect(mock.find(id)?.stepState).toEqual({ a: "user-1" });

    forceFail = false;
    await new Promise((r) => setTimeout(r, 10));
    await worker.processOnce();
    expect(mock.find(id)?.status).toBe("succeeded");
    expect(runs.a).toBe(1);
    expect(runs.b).toBe(2);
  });

  test("runLoop processes runs concurrently up to concurrency limit", async () => {
    const total = 5;
    let peak = 0;
    let inFlight = 0;
    let finished = 0;
    let resolveAllDone: () => void;
    const allDone = new Promise<void>((r) => (resolveAllDone = r));

    const worker = new Worker({
      client,
      workerId: "w1",
      concurrency: total,
      pollIntervalMs: 5,
      idleTimeoutMs: 500,
      registry: registry({
        slow: async () => {
          inFlight++;
          peak = Math.max(peak, inFlight);
          await new Promise((r) => setTimeout(r, 50));
          inFlight--;
          finished++;
          if (finished === total) resolveAllDone();
          return "ok";
        },
      }),
    });

    const ids: string[] = [];
    for (let i = 0; i < total; i++) {
      ids.push(mock.seed({ name: "slow", payload: { i } }));
    }

    worker.start();
    await allDone;
    await worker.drain();

    expect(peak).toBeGreaterThanOrEqual(2);
    for (const id of ids) {
      expect(mock.find(id)?.status).toBe("succeeded");
    }
  });
});

File: sdk/go/worker_test.go
package tako

import (
	"context"
	"encoding/json"
	"errors"
	"net"
	"os"
	"path/filepath"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

// MockServer emulates the tako-server enqueue socket with an in-memory
// task store — just enough to drive worker lifecycle tests end-to-end.
type mockServer struct {
	l        net.Listener
	path     string
	mu       sync.Mutex
	tasks    map[string]*Run
	idCount  int
	received []map[string]any
}

func startMockServer(t *testing.T) *mockServer {
	t.Helper()
	dir, err := os.MkdirTemp("/tmp", "tako-worker-mock-")
	if err != nil {
		t.Fatalf("tempdir: %v", err)
	}
	t.Cleanup(func() { _ = os.RemoveAll(dir) })
	path := filepath.Join(dir, "srv.sock")
	l, err := net.Listen("unix", path)
	if err != nil {
		t.Fatalf("listen: %v", err)
	}
	s := &mockServer{l: l, path: path, tasks: map[string]*Run{}}
	go s.accept()
	t.Cleanup(func() { _ = l.Close() })
	return s
}

func (s *mockServer) accept() {
	for {
		conn, err := s.l.Accept()
		if err != nil {
			return
		}
		go s.handleConn(conn)
	}
}

func (s *mockServer) handleConn(c net.Conn) {
	defer c.Close()
	buf := make([]byte, 0, 1024)
	tmp := make([]byte, 1024)
	for {
		n, err := c.Read(tmp)
		if n > 0 {
			buf = append(buf, tmp[:n]...)
			for {
				i := indexOfByte(buf, '\n')
				if i == -1 {
					break
				}
				line := buf[:i]
				buf = buf[i+1:]
				var cmd map[string]any
				if err := json.Unmarshal(line, &cmd); err != nil {
					c.Write([]byte(`{"status":"error","message":"bad json"}` + "\n"))
					continue
				}
				s.mu.Lock()
				s.received = append(s.received, cmd)
				s.mu.Unlock()
				resp := s.dispatch(cmd)
				out, _ := json.Marshal(resp)
				c.Write(append(out, '\n'))
			}
		}
		if err != nil {
			return
		}
	}
}

func indexOfByte(buf []byte, b byte) int {
	for i, v := range buf {
		if v == b {
			return i
		}
	}
	return -1
}

func (s *mockServer) seed(name string, maxAttempts uint32) string {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.idCount++
	id := "t" + itoa(s.idCount)
	s.tasks[id] = &Run{
		ID:          id,
		Name:        name,
		Payload:     json.RawMessage("{}"),
		Status:      "pending",
		Attempts:    0,
		MaxAttempts: maxAttempts,
		RunAtMs:     time.Now().UnixMilli(),
		StepState:   map[string]any{},
	}
	return id
}

func (s *mockServer) find(id string) *Run {
	s.mu.Lock()
	defer s.mu.Unlock()
	t := s.tasks[id]
	return t
}

func (s *mockServer) dispatch(cmd map[string]any) map[string]any {
	s.mu.Lock()
	defer s.mu.Unlock()

	switch cmd["command"] {
	case "claim_run":
		rawNames, _ := cmd["names"].([]any)
		names := make([]string, 0, len(rawNames))
		for _, v := range rawNames {
			if str, ok := v.(string); ok {
				names = append(names, str)
			}
		}
		for _, t := range s.tasks {
			if t.Status == "pending" && contains(names, t.Name) && t.RunAtMs <= time.Now().UnixMilli() {
				t.Status = "running"
				t.Attempts++
				return map[string]any{
					"status": "ok",
					"data": map[string]any{
						"id":           t.ID,
						"name":         t.Name,
						"payload":      json.RawMessage(t.Payload),
						"status":       t.Status,
						"attempts":     t.Attempts,
						"max_attempts": t.MaxAttempts,
						"run_at_ms":    t.RunAtMs,
						"step_state":   t.StepState,
					},
				}
			}
		}
		return map[string]any{"status": "ok", "data": nil}
	case "heartbeat_run":
		return map[string]any{"status": "ok", "data": map[string]any{}}
	case "save_step":
		id, _ := cmd["id"].(string)
		stepName, _ := cmd["step_name"].(string)
		if t := s.tasks[id]; t != nil {
			if t.StepState == nil {
				t.StepState = map[string]any{}
			}
			t.StepState[stepName] = cmd["result"]
		}
		return map[string]any{"status": "ok", "data": map[string]any{}}
	case "complete_run":
		id, _ := cmd["id"].(string)
		if t := s.tasks[id]; t != nil {
			t.Status = "succeeded"
		}
		return map[string]any{"status": "ok", "data": map[string]any{}}
	case "cancel_run":
		id, _ := cmd["id"].(string)
		if t := s.tasks[id]; t != nil {
			t.Status = "cancelled"
		}
		return map[string]any{"status": "ok", "data": map[string]any{}}
	case "defer_run":
		id, _ := cmd["id"].(string)
		if t := s.tasks[id]; t != nil {
			t.Status = "pending"
			if w, ok := cmd["wake_at_ms"].(float64); ok {
				t.RunAtMs = int64(w)
			}
		}
		return map[string]any{"status": "ok", "data": map[string]any{}}
	case "wait_for_event":
		id, _ := cmd["id"].(string)
		if t := s.tasks[id]; t != nil {
			t.Status = "pending"
		}
		return map[string]any{"status": "ok", "data": map[string]any{}}
	case "fail_run":
		id, _ := cmd["id"].(string)
		finalize, _ := cmd["finalize"].(bool)
		if t := s.tasks[id]; t != nil {
			if finalize {
				t.Status = "dead"
			} else {
				t.Status = "pending"
				if next, ok := cmd["next_run_at_ms"].(float64); ok {
					t.RunAtMs = int64(next)
				}
			}
		}
		return map[string]any{"status": "ok", "data": map[string]any{}}
	case "register_schedules":
		return map[string]any{"status": "ok", "data": map[string]any{}}
	case "enqueue_run":
		s.idCount++
		id := "e" + itoa(s.idCount)
		return map[string]any{
			"status": "ok",
			"data":   map[string]any{"id": id, "deduplicated": false},
		}
	default:
		return map[string]any{"status": "error", "message": "unknown"}
	}
}

func contains(xs []string, s string) bool {
	for _, v := range xs {
		if v == s {
			return true
		}
	}
	return false
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	buf := [20]byte{}
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	return string(buf[i:])
}

// Tests

func resetRegistry() {
	registryMu.Lock()
	registry = map[string]workflowRegistration{}
	registryStarted = false
	registryMu.Unlock()
}

func TestEnqueueSendsEnqueueTaskCommand(t *testing.T) {
	s := startMockServer(t)
	t.Setenv("TAKO_INTERNAL_SOCKET", s.path)
	t.Setenv("TAKO_APP_NAME", "test-app")
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	s.seed("send-email", 3)
	r, err := Enqueue(ctx, "send-email", map[string]any{"to": "a@b.c"}, EnqueueOpts{})
	if err != nil {
		t.Fatalf("enqueue: %v", err)
	}
	if r.ID == "" {
		t.Fatalf("empty id")
	}
}

func TestEnqueueSerializesOpts(t *testing.T) {
	s := startMockServer(t)
	t.Setenv("TAKO_INTERNAL_SOCKET", s.path)
	t.Setenv("TAKO_APP_NAME", "test-app")
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	when := time.UnixMilli(1_700_000_000_000)
	max := uint32(5)
	key := "cron:5m:0"
	_, _ = Enqueue(ctx, "w", nil, EnqueueOpts{RunAt: &when, MaxAttempts: &max, UniqueKey: &key})

	s.mu.Lock()
	defer s.mu.Unlock()
	opts, _ := s.received[0]["opts"].(map[string]any)
	if opts["run_at_ms"].(float64) != float64(when.UnixMilli()) {
		t.Fatalf("run_at_ms: %v", opts)
	}
	if opts["max_attempts"].(float64) != 5 {
		t.Fatalf("max_attempts: %v", opts)
	}
	if opts["unique_key"].(string) != "cron:5m:0" {
		t.Fatalf("unique_key: %v", opts)
	}
}

func TestWorkerRunsHandlerAndCompletes(t *testing.T) {
	resetRegistry()
	s := startMockServer(t)
	t.Setenv("TAKO_INTERNAL_SOCKET", s.path)
	t.Setenv("TAKO_APP_NAME", "test-app")

	called := atomic.Bool{}
	RegisterWorkflow("echo", func(ctx *WorkflowContext, p json.RawMessage) error {
		called.Store(true)
		return nil
	})

	id := s.seed("echo", 3)

	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	done := make(chan struct{})
	go func() {
		_ = RunWorker(ctx)
		close(done)
	}()

	deadline := time.Now().Add(3 * time.Second)
	for time.Now().Before(deadline) {
		if called.Load() && s.find(id).Status == "succeeded" {
			cancel()
			break
		}
		time.Sleep(20 * time.Millisecond)
	}
	<-done

	if !called.Load() {
		t.Fatalf("handler never ran")
	}
	if s.find(id).Status != "succeeded" {
		t.Fatalf("task not succeeded: %v", s.find(id).Status)
	}
}

func TestStepRunMemoizesAcrossRetries(t *testing.T) {
	resetRegistry()
	s := startMockServer(t)
	t.Setenv("TAKO_INTERNAL_SOCKET", s.path)
	t.Setenv("TAKO_APP_NAME", "test-app")

	var aRuns, bRuns atomic.Uint32
	forceFail := atomic.Bool{}
	forceFail.Store(true)

	RegisterWorkflow("multi", func(ctx *WorkflowContext, p json.RawMessage) error {
		var got string
		if err := ctx.Step.Run("a", &got, func() (any, error) {
			aRuns.Add(1)
			return "user-1", nil
		}); err != nil {
			return err
		}
		return ctx.Step.Run("b", nil, func() (any, error) {
			bRuns.Add(1)
			if forceFail.Load() {
				return nil, errors.New("fail-b")
			}
			return got, nil
		})
	})

	s.seed("multi", 5)

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	done := make(chan struct{})
	go func() {
		_ = RunWorker(ctx)
		close(done)
	}()

	// First attempt runs + fails on b.
	time.Sleep(600 * time.Millisecond)
	forceFail.Store(false)
	time.Sleep(2 * time.Second)
	cancel()
	<-done

	if aRuns.Load() != 1 {
		t.Fatalf("step a should have run exactly once, got %d", aRuns.Load())
	}
	if bRuns.Load() < 2 {
		t.Fatalf("step b should have run at least twice, got %d", bRuns.Load())
	}
}
